### ISSUE TITLE
slow speeds and "smb2_ioctl" ?

### ISSUE THREAD

**Issue by Andy2244** [cifsd-team/cifsd#174]
Just forwarding this, some users report slow speeds on hardware that should easily do much more. 
_(Aka 14MB/s on hardware that can do 200MB/s)_ 
They see this in logs? So could this affect speeds and if so what can we do about it?

```
kern.err kernel: [51215.420589] kcifsd: smb2_ioctl:6584: br-lan speed is unknown, defaulting to 100
kern.err kernel: [51215.427970] kcifsd: smb2_ioctl:6584: wg0 speed is unknown, defaulting to 100
```

**Comment by sergey-senozhatsky** [comment]
On (07/29/19 05:42), Andy Walsh wrote:
>    Just forwarding this, some users report slow speeds on hardware that
>    should easily do much more. They see this in logs? So could this affect
>    speeds and if so what can we do about it?
> 
>  kern.err kernel: [51215.420589] kcifsd: smb2_ioctl:6584: br-lan speed is unknown, defaulting to 100
>  kern.err kernel: [51215.427970] kcifsd: smb2_ioctl:6584: wg0 speed is unknown, defaulting to 100

Hmm. Do they have common H/W?

If we change the default value

---
diff --git a/smb2pdu.c b/smb2pdu.c
index b7b362e..1a0ad20 100644
--- a/smb2pdu.c
+++ b/smb2pdu.c
@@ -6297,7 +6297,7 @@ int smb2_ioctl(struct cifsd_work *work)
 			else {
 				cifsd_err("%s speed is unknown, defaulting to 100\n",
 					netdev->name);
-				speed = 1000;
+				speed = SPEED_1000;
 			}
 
 			nii_rsp->LinkSpeed = cpu_to_le64(speed * 1000000);
---

Does it get any better?

Other default value might be SPEED_10000

 SPEED_10000		"10Gb/sec"
 SPEED_1000		"1Gb/sec"
 SPEED_100		"100Mbit/sec"

	-ss

**Comment by Andy2244** [comment]
> Do they have common H/W?

From what i can guess via shell name its on a ZyXEL - NSA310

**Comment by kofec** [comment]
Hi, 
I reported on Openwrt forum. It is possible to change to  "1Gb/sec" by default ? Can I ask why do you detect interface speed (you can optimized utilization base on it )?

**Comment by sergey-senozhatsky** [comment]
On (07/29/19 05:57), Andy Walsh wrote:
>      Do they have common H/W?
> 
>    From what i can guess via shell name its on a ZyXEL - NSA310

Uhh... The error message is misleading, it says 100:

	cifsd_err("%s speed is unknown, defaulting to 100\n")

but kcifsd actually sets speed to 1000

	speed = 1000;

	-ss

**Comment by namjaejeon** [comment]
I would like to know whether it is related with SMB3 network inteface info ioctl first.

Could you please check speed after adding max protocol parameter in smb.conf ?
It limits connection to above SMB3.X.

Thanks!

**Comment by sergey-senozhatsky** [comment]
> It is possible to change to "1Gb/sec" by default ?

Yeah, that's what we do. The error message is inaccurate.

-ss

**Comment by kofec** [comment]
@namjaejeon 
I don't see this sentence "kern.err kernel: .... kcifsd: smb2_ioct ..." when change my config to:

```
root@NSA310:~# cat /etc/cifs/smb.conf
[global]
        netbios name = NSA310
        server string = Cifsd on OpenWrt
        workgroup = OpenWrt
        interfaces = br-lan
        bind interfaces only = yes
        ipc timeout = 8
        deadtime = 15
        map to guest = Bad User
        server max protocol = SMB2

[dysk3TB]
        path = /mnt/sda1
        force user = root
        force group = root
        browseable = yes
        read only = no
        writeable = yes
        guest ok = yes
```
I open new issue not to mix with this one for "server max protocol = SMB1"
https://github.com/cifsd-team/cifsd/issues/176

**Comment by namjaejeon** [comment]
Okay, good.
performance ?

**Comment by sergey-senozhatsky** [comment]
AFAICT, performance is still around 14 MB/s for smb2.1 and smb3.


> 14MB/s on hardware that can do 200MB/s

@Andy2244 Where does 200 MB/s come from? Does samba demonstrate 200 MB/s?

-ss

**Comment by Andy2244** [comment]
> Where does 200 MB/s come from? Does samba demonstrate 200 MB/s?

Nah in the org. post was a DD check, that showed the disk could do way more than the 14 MB/s, so it should be limited by network speed not disk/interface speed.

**Comment by sergey-senozhatsky** [comment]
> > Where does 200 MB/s come from? Does samba demonstrate 200 MB/s?
> 
> Nah in the org. post was a DD check, that showed the disk could do way more than the 14 MB/s, so it should be limited by network speed not disk/interface speed. Was just a check, because sometimes router usb/ata interfaces are crippled/bugged.

Well, we can't compete with `dd` :) Does samba show a significantly faster I/O than 14 MB/s?

-ss

**Comment by Andy2244** [comment]
> Does samba show a significantly faster I/O than 14 MB/s?

The post is over [here](https://forum.openwrt.org/t/slow-samba-nfs-read-in-openwrt-stable-dev/40366/46), i'm also a little confused at this point. There are two users that have general speed issues with all file-services (nfs/samba/cifsd) yet the latest post indicates that there are now two users reporting this 14 MB/s number....

**Comment by Andy2244** [comment]
Will just close this, just update the debug text and maybe investigate why it cant figure out the speed for the bridge/interface and falls back to defaults.

**Comment by sergey-senozhatsky** [comment]
> > Does samba show a significantly faster I/O than 14 MB/s?
> 
> The post is over [here](https://forum.openwrt.org/t/slow-samba-nfs-read-in-openwrt-stable-dev/40366/46), i'm also a little confused at this point. There are two users that have general speed issues with all file-services (nfs/samba/cifsd) yet the latest post indicates that there are now two users reporting this 14 MB/s number for cifsd ....

Hmm. Would be a bit easier if we could compare samba and cifsd. Right now I'm not sure if 14 MB/s
is really slow or ok-ish.

BTW, we have a patch in cifsd-next (I think it's not in the master yet) which increases MaxRead/Write
buffer size for smb2/smb3. This should reduce the number of network packets and latencies and
increase I/O speed.

-ss

**Comment by sergey-senozhatsky** [comment]
> BTW, we have a patch in cifsd-next (I think it's not in the master yet) which increases MaxRead/Write
> buffer size for smb2/smb3. This should reduce the number of network packets and latencies and
> increase I/O speed.

Hmm... actually need to check that patch more. The problem is that socket recv/send buffer size is 65K.
So in this case we are not really making any difference in terms of number of packets.

@namjaejeon how did you test that patch? I'm not exactly sure if the test restuls were not accidental.

-ss

**Comment by namjaejeon** [comment]
@sergey-senozhatsky  yes. I checked wireshark and I saw data size is 4MB in smb request. 1 to 13% is measured by steve french in cifs. I have seen 2% performance improvement in my case. And  tcp stack may split smb request as per segment and send out.  The meaning of 4MB of smb request is not that. 1MB vs 4MB request means smb header size of request and response  are decreased in total packets. So performance improvement is very small. I am not sure how steve get 13% improvement. maybe hardware difference ?
